Failsafe TestHTMLReporter parameter.toString

### TestNG Version
7.6.2-SNAPSHOT

### Expected behavior
Failsafe `TestHTMLReporter` when parameter.toString() of test execution throws RuntimeException.

### Actual behavior
`TestHTMLReporter` does not fail if test parameter toString method throws Exception. Sequentially it leads to maven build failure with non-clear diagnostics (missing full stacktrace, only message) like
```
[INFO] Results:
[INFO] 
[INFO] Tests run: 0, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
...
build	10-Nov-2022 05:44:17	[INFO] ------------------------------------------------------------------------
build	10-Nov-2022 05:44:17	[INFO] BUILD FAILURE
build	10-Nov-2022 05:44:17	[INFO] ------------------------------------------------------------------------
build	10-Nov-2022 05:44:17	[INFO] Total time:  10:19 min (Wall Clock)
build	10-Nov-2022 05:44:17	[INFO] Finished at: 2022-11-10T05:44:17Z
build	10-Nov-2022 05:44:17	[INFO] ------------------------------------------------------------------------
build	10-Nov-2022 05:44:17	[ERROR] Failed to execute goal org.apache.maven.plugins:maven-failsafe-plugin:2.22.2:verify (default-cli) on project users-integration-test: There are test failures.
build	10-Nov-2022 05:44:17	[ERROR] 
build	10-Nov-2022 05:44:17	[ERROR] Please refer to /home/bamboo/bamboo-agent-home/xml-data/build-dir/RTB-NRB22926-JOB10/server/general/users-integration-test/target/failsafe-reports for the individual test results.
build	10-Nov-2022 05:44:17	[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
build	10-Nov-2022 05:44:17	[ERROR] org.apache.maven.surefire.booter.SurefireBooterForkException: There was an error in the forked process
build	10-Nov-2022 05:44:17	[ERROR] game_type_id=10000006, model=Model [class=projects.pt.server.users.models.userid.UserIdModel, name=UserId, key=-4268742994815057915:UserId, id=145, serverOnly=true], field=PersistentFieldPersistentModelField[UserIdEntity UserIdModel.userIdEntity]
build	10-Nov-2022 05:44:17	[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.fork(ForkStarter.java:656)
build	10-Nov-2022 05:44:17	[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:282)
build	10-Nov-2022 05:44:17	[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:245)
build	10-Nov-2022 05:44:17	[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeProvider(AbstractSurefireMojo.java:1183)
build	10-Nov-2022 05:44:17	[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeAfterPreconditionsChecked(AbstractSurefireMojo.java:1011)
build	10-Nov-2022 05:44:17	[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.execute(AbstractSurefireMojo.java:857)
build	10-Nov-2022 05:44:17	[ERROR] 	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
build	10-Nov-2022 05:44:17	[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:210)
build	10-Nov-2022 05:44:17	[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156)
build	10-Nov-2022 05:44:17	[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148)
build	10-Nov-2022 05:44:17	[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
build	10-Nov-2022 05:44:17	[ERROR] 	at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call(MultiThreadedBuilder.java:190)
build	10-Nov-2022 05:44:17	[ERROR] 	at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call(MultiThreadedBuilder.java:186)
build	10-Nov-2022 05:44:17	[ERROR] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
build	10-Nov-2022 05:44:17	[ERROR] 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
build	10-Nov-2022 05:44:17	[ERROR] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
build	10-Nov-2022 05:44:17	[ERROR] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
build	10-Nov-2022 05:44:17	[ERROR] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
build	10-Nov-2022 05:44:17	[ERROR] 	at java.base/java.lang.Thread.run(Thread.java:833)
build	10-Nov-2022 05:44:17	[ERROR] -> [Help 1]
build	10-Nov-2022 05:44:17	org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-failsafe-plugin:2.22.2:verify (default-cli) on project users-integration-test: There are test failures.
```
The good thing is that failure can be found in `target/failsafe-reports/*-jvmRun*.dump` and there is stacktrace. But if test suite is executed on the CI/CD it can be problem to obtain it.
How it can happen: objects that are parameters of tests may have stateful `toString()` method (actually antipattern, but failsafe still makes sense to handle such failures).

### Is the issue reproducible on runner?

- [ ] Shell
- [x] Maven (with HTML reporting)
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [ ] IntelliJ
- [ ] NetBeans


### Test case sample
The PR with the proposed fix https://github.com/cbeust/testng/pull/2831

### Contribution guidelines

Incase you plan to raise a pull request to fix this issue, please make sure you refer our [Contributing](.github/CONTRIBUTING.md) section for detailed set of steps.

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -9,7 +9,8 @@ Fixed: GITHUB-2800: Running Test Classes with Inherited @Factory and @DataProvid
 New: Ability to provide custom error message for assertThrows\expectThrows methods (Anatolii Yuzhakov)
 Fixed: GITHUB-2780: Use SpotBugs instead of abandoned FindBugs
 Fixed: GITHUB-2801: JUnitReportReporter is too slow
-Fixed: GITHUB-2807: buildStackTrace should be fail-safe
+Fixed: GITHUB-2807: buildStackTrace should be fail-safe (Sergey Chernov)
+Fixed: GITHUB-2830: TestHTMLReporter parameter toString should be fail-safe (Sergey Chernov)
 Fixed: GITHUB-2798: Parallel executions coupled with retry analyzer results in duplicate retry analyzer instances being created (Krishnan Mahadevan)
 
 7.6.1

--- a/testng-core/src/main/java/org/testng/reporters/TestHTMLReporter.java
+++ b/testng-core/src/main/java/org/testng/reporters/TestHTMLReporter.java
@@ -123,7 +123,20 @@ public class TestHTMLReporter implements ITestListener {
           if (j > 0) {
             pw.append(", ");
           }
-          pw.append(parameters[j] == null ? "null" : parameters[j].toString());
+          if (parameters[j] == null) {
+            pw.append("null");
+          } else {
+            String parameterToString;
+            try {
+              parameterToString = parameters[j].toString();
+            } catch (RuntimeException | Error e) {
+              log(e.toString());
+              // failover in case parameter toString() cannot be evaluated
+              parameterToString =
+                  parameters[j].getClass().getName() + "@" + System.identityHashCode(parameters[j]);
+            }
+            pw.append(parameterToString);
+          }
         }
       }
 

--- a/testng-core/src/test/java/org/testng/reporters/TestHTMLReporterTest.java
+++ b/testng-core/src/test/java/org/testng/reporters/TestHTMLReporterTest.java
@@ -1,0 +1,49 @@
+package org.testng.reporters;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.Collections;
+import java.util.List;
+import org.testng.ITestClass;
+import org.testng.ITestNGMethod;
+import org.testng.ITestResult;
+import org.testng.annotations.Test;
+import org.testng.internal.TestResult;
+
+public class TestHTMLReporterTest {
+
+  @Test(description = "GITHUB-2830")
+  public void generateTableParametersToStringShouldBeFailsafe() {
+    ITestClass testClass = mock(ITestClass.class);
+    when(testClass.getName()).thenReturn("testClass");
+
+    ITestNGMethod testNGMethod = mock(ITestNGMethod.class);
+    when(testNGMethod.getMethodName()).thenReturn("testMethod");
+    when(testNGMethod.getTestClass()).thenReturn(testClass);
+
+    TestResult testResult = TestResult.newEmptyTestResult();
+    testResult.setMethod(testNGMethod);
+
+    testResult.setParameters(new Object[] {new ThrowingOnToString()});
+
+    List<ITestResult> tests = Collections.singletonList(testResult);
+
+    StringWriter sw = new StringWriter();
+    PrintWriter pw = new PrintWriter(sw, true);
+    TestHTMLReporter.generateTable(pw, "title", tests, "cssClass", (t1, t2) -> 0);
+
+    assertThat(sw.toString())
+        .contains("Parameters: org.testng.reporters.TestHTMLReporterTest$ThrowingOnToString@");
+  }
+
+  private static class ThrowingOnToString {
+    @Override
+    public String toString() {
+      throw new IllegalStateException("Cannot calculate toString");
+    }
+  }
+}

